cluster/images/etcd/migrate-if-needed.sh

```diff
@@ -34,22 +34,56 @@ if [ -z "${DATA_DIRECTORY:-}" ]; then
   exit 0
 fi
 
+ETCD="${ETCD:-/usr/local/bin/etcd}"
 ETCDCTL="${ETCDCTL:-/usr/local/bin/etcdctl}"
 VERSION_FILE="version.txt"
 CURRENT_STORAGE='etcd2'
 if [ -e "${DATA_DIRECTORY}/${VERSION_FILE}" ]; then
   CURRENT_STORAGE="$(cat ${DATA_DIRECTORY}/${VERSION_FILE})"
 fi
 
+start_etcd() {
+  ETCD_PORT=18629
+  ETCD_PEER_PORT=18630
+  ${ETCD} --data-dir=${DATA_DIRECTORY} \
+    --listen-client-urls http://127.0.0.1:${ETCD_PORT} \
+    --advertise-client-urls http://127.0.0.1:${ETCD_PORT} \
+    --listen-peer-urls http://127.0.0.1:${ETCD_PEER_PORT} \
+    --initial-advertise-peer-urls http://127.0.01:${ETCD_PEER_PORT} \
+    1>>/dev/null 2>&1 &
+  ETCD_PID=$!
+  # TODO: Wait until etcd is up.
+  sleep 5
+}
+
+stop_etcd() {
+  kill "${ETCD_PID-}" >/dev/null 2>&1 || :
+  wait "${ETCD_PID-}" >/dev/null 2>&1 || :
+}
+
 if [ "${CURRENT_STORAGE}" = "etcd2" -a "${TARGET_STORAGE}" = "etcd3" ]; then
   # If directory doesn't exist or is empty, this means that there aren't any
   # data for migration, which means we can skip this step.
   if [ -d "${DATA_DIRECTORY}" ]; then
     if [ "$(ls -A ${DATA_DIRECTORY})" ]; then
       echo "Performing etcd2 -> etcd3 migration"
-      # TODO: Pass a correct transformer to handle TTLs.
-      echo "ETCDCTL_API=3 ${ETCDCTL} migrate --data-dir=${DATA_DIRECTORY}"
       ETCDCTL_API=3 ${ETCDCTL} migrate --data-dir=${DATA_DIRECTORY}
+      echo "Attaching leases to TTL entries"
+      # Now attach lease to all keys.
+      # To do it, we temporarily start etcd on a random port (so that
+      # apiserver actually cannot access it).
+      start_etcd
+      # List all keys that require attaching a lease.
+      # Since we still didn't remove v2 data, we can simply use it.
+      ${ETCDCTL} --endpoints=127.0.0.1:${ETCD_PORT} ls --recursive \
+        "${TTL_KEYS_DIRECTORY:-/registry/events}" > ttl_keys.txt
```

Comment:
@xiang90 @wojtek-t 
Since we already assumed the prefix is known here as "TTL_KEYS_DIRECTORY", we can save this step of listing ttl keys.
Instead, we can directly list prefix in "attachlease" program.